Modules/_heapqmodule.c

```diff
@@ -538,6 +538,460 @@ _heapq__heapify_max(PyObject *module, PyObject *heap)
+                        arr[pos] = NULL;
+                        arr[pos + 1] = NULL;
+                        return -1;
+                    }
+                    if (cmp > 0) {
+                        childpos += 2;
+                        childkey = otherchildkey;
+                        childvalue = otherchildvalue;
+                    }
+                }
+            }
+            arr[pos] = childkey;
+            arr[pos + 1] = childvalue;
+            pos = childpos;
+        }
+    }
+    arr[pos] = NULL;
+    arr[pos + 1] = NULL;
+
+    it = PyList_GET_ITEM(mo->iterators, pos/2 - (n - 1));
+    if (it == NULL)
+        return 0;
+    childvalue = PyIter_Next(it);
+    if (childvalue == NULL) {
+        Py_DECREF(it);
+        PyList_SET_ITEM(mo->iterators, pos/2 - (n - 1), NULL);
+        if (PyErr_Occurred())
+            return -1;
+    }
+    else {
+        childkey = _PyObject_CallOneArg(mo->keyfunc, childvalue);
+        if (childkey == NULL) {
+            Py_DECREF(childvalue);
+            return -1;
+        }
+        arr[pos] = childkey;
+        arr[pos+1] = childvalue;
+    }
+    return 0;
+}
+
+/* shift = 2 * n - (1 << n.bit_length()) */
+static Py_ssize_t
+get_iter_shift(Py_ssize_t n) {
+    int n0 = n;
+    int m = 1;
+    while (n > 0) {
+        n >>= 1;
+        m <<= 1;
+    }
+    return (2 * n0) - m;
+}
+
+static PyObject *
+merge_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyObject *keyfunc = NULL;
+    int reverse = 0;
+    mergeobject *mo;
+    PyObject *iterators = NULL, *tree = NULL;
+    Py_ssize_t num_iters, shift, i;
+
+    if (kwds != NULL) {
+        char *kwlist[] = {"key", "reverse", 0};
+        PyObject *tmpargs = PyTuple_New(0);
+        if (tmpargs == NULL)
+            return NULL;
+        if (!PyArg_ParseTupleAndKeywords(tmpargs, kwds, "|Op:merge",
+                                         kwlist, &keyfunc, &reverse)) {
+            Py_DECREF(tmpargs);
+            return NULL;
+        }
+        Py_DECREF(tmpargs);
+    }
+
+    if (keyfunc == Py_None)
+        keyfunc = NULL;
+    Py_XINCREF(keyfunc);
+
+    if (keyfunc != NULL && !PyCallable_Check(keyfunc)) {
+        PyErr_SetString(PyExc_TypeError, "Key must be callable or None.");
+        goto error;
+    }
+
+    assert(PyTuple_CheckExact(args));
+    num_iters = PyTuple_GET_SIZE(args);
+
+    if (num_iters == 0) {
+        mo = (mergeobject *)type->tp_alloc(type, 0);
+        if (mo == NULL)
+            return NULL;
+        mo->iterators = NULL;
+        mo->tree = NULL;
+        mo->keyfunc = NULL;
+        mo->reverse = 0;
+        mo->started = 1;
+        return (PyObject *) mo;
+    }
+
+    iterators = PyList_New(num_iters);
+    if (iterators == NULL)
+        goto error;
+
+    /* Store the tree as a list with n-1 internal nodes and n leaves.
+       If a key is being used, store keys next to original items. */
+    tree = PyList_New(keyfunc? num_iters * 4 - 2: num_iters * 2 - 1);
```

Comment:
This list is left uninitialized but is tracked by the collector and therefore it can crash (also if is ever exposed to Python). Notice that the `PyList_New` docs mention this:

> Note If len is greater than zero, the returned list object’s items are set to NULL. Thus you cannot use abstract API functions such as PySequence_SetItem() or expose the object to Python code before setting all items to a real object with PyList_SetItem().